### ISSUE TITLE
build fails with libressl 2.7

### ISSUE THREAD

**Issue by ncopa** [nghttp2/nghttp2#1156]
libressl 2.7 implemented some of the [openssl 1.1 API](https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-2.7.2-relnotes.txt).

This makes build fail:
````
  CXX      libnghttpx_a-shrpx_connection.o
shrpx_connection.cc: In function 'int shrpx::{anonymous}::shrpx_bio_write(BIO*, const char*, int)':
shrpx_connection.cc:159:55: error: call of overloaded 'BIO_get_data(BIO*&)' is ambiguous
   auto conn = static_cast<Connection *>(BIO_get_data(b));
                                                       ^
shrpx_connection.cc:49:7: note: candidate: void* shrpx::BIO_get_data(BIO*)
 void *BIO_get_data(BIO *bio) { return bio->ptr; }
       ^~~~~~~~~~~~
In file included from /usr/include/openssl/evp.h:67:0,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from shrpx_config.h:47,
                 from shrpx_connection.h:28,
                 from shrpx_connection.cc:25:
/usr/include/openssl/bio.h:608:8: note: candidate: void* BIO_get_data(BIO*)
 void  *BIO_get_data(BIO *a);
        ^~~~~~~~~~~~
shrpx_connection.cc: In function 'int shrpx::{anonymous}::shrpx_bio_read(BIO*, char*, int)':
shrpx_connection.cc:200:55: error: call of overloaded 'BIO_get_data(BIO*&)' is ambiguous
   auto conn = static_cast<Connection *>(BIO_get_data(b));
                                                       ^
shrpx_connection.cc:49:7: note: candidate: void* shrpx::BIO_get_data(BIO*)
 void *BIO_get_data(BIO *bio) { return bio->ptr; }
       ^~~~~~~~~~~~
In file included from /usr/include/openssl/evp.h:67:0,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from shrpx_config.h:47,
                 from shrpx_connection.h:28,
                 from shrpx_connection.cc:25:
/usr/include/openssl/bio.h:608:8: note: candidate: void* BIO_get_data(BIO*)
 void  *BIO_get_data(BIO *a);
        ^~~~~~~~~~~~
shrpx_connection.cc: In member function 'void shrpx::Connection::set_ssl(SSL*)':
shrpx_connection.cc:311:25: error: call of overloaded 'BIO_set_data(bio_st*&, shrpx::Connection*)' is ambiguous
   BIO_set_data(bio, this);
                         ^
shrpx_connection.cc:50:6: note: candidate: void shrpx::BIO_set_data(BIO*, void*)
 void BIO_set_data(BIO *bio, void *ptr) { bio->ptr = ptr; }
      ^~~~~~~~~~~~
In file included from /usr/include/openssl/evp.h:67:0,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from shrpx_config.h:47,
                 from shrpx_connection.h:28,
                 from shrpx_connection.cc:25:
/usr/include/openssl/bio.h:609:7: note: candidate: void BIO_set_data(BIO*, void*)
 void  BIO_set_data(BIO *a, void *ptr);
       ^~~~~~~~~~~~
make[3]: *** [Makefile:1909: libnghttpx_a-shrpx_connection.o] Error 1
````

I think the proper solution would be to have configure test for `BIO_get_data`, `BIO_set_data`, and `BIO_set_init`

**Comment by ncopa** [comment]
Something like this:
````diff
diff --git a/configure.ac b/configure.ac
index 9445f93..1a8eae5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -359,6 +359,11 @@ PKG_CHECK_MODULES([OPENSSL], [openssl >= 1.0.1],
 if test "x${have_openssl}" = "xno"; then
   AC_MSG_NOTICE($OPENSSL_PKG_ERRORS)
 fi
+save_LIBS=$LIBS
+LIBS=$OPENSSL_LIBS
+AC_CHECK_FUNCS(BIO_set_data BIO_get_data BIO_set_init)
+LIBS=$save_LIBS
+
 
 # c-ares (for src)
 PKG_CHECK_MODULES([LIBCARES], [libcares >= 1.7.5], [have_libcares=yes],
diff --git a/src/shrpx_connection.cc b/src/shrpx_connection.cc
index 06ad958..e5cf1be 100644
--- a/src/shrpx_connection.cc
+++ b/src/shrpx_connection.cc
@@ -44,13 +44,15 @@ using namespace nghttp2;
 
 namespace shrpx {
 
-#if !OPENSSL_1_1_API
-
+#ifndef HAVE_BIO_GET_DATA
 void *BIO_get_data(BIO *bio) { return bio->ptr; }
+#endif
+#ifndef HAVE_BIO_SET_DATA
 void BIO_set_data(BIO *bio, void *ptr) { bio->ptr = ptr; }
+#endif
+#ifndef HAVE_BIO_SET_INIT
 void BIO_set_init(BIO *bio, int init) { bio->init = init; }
-
-#endif // !OPENSSL_1_1_API
+#endif
 
 Connection::Connection(struct ev_loop *loop, int fd, SSL *ssl,
                        MemchunkPool *mcpool, ev_tstamp write_timeout,
````

**Comment by tatsuhiro-t** [comment]
#1162 fixed libressl 2.7 issue.
